resources/widgetlist.ts

```diff
@@ -21,41 +40,56 @@ class WidgetList extends HTMLElement {
   // and the second being the direction is wraps for the next
   // row/column. eg. "left down" will grow a list toward the left,
   // and subsequent rows will be below the first.
-  set toward(s) {
-    this.setAttribute('toward', s);
+  set toward(s: string | null) {
+    if (s === null)
+      this.removeAttribute('toward');
+    else
+      this.setAttribute('toward', s);
```

Comment:
I'm not sure I understand these changes to the setters.  Is it possible to have `set toward(s: string)` with no null possibility here, and elsewhere? The `removeAttribute` does not seem like how it behaved previously?

Reply:
originally, it's `this.setAttribute('toward', w);`, which according to MDN, `setAttribute('prop', null)` means `removeAttribute`

Reply:
Can you link me to this documentation?

I guess my confusion here is that we are providing this `toward` API here, and so we can assert "you must pass a non-null string".  I don't think any caller of any of these components in resources would want to remove an attribute, only set the value.

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute#parameters

Reply:
Seems like I should change like this in my PR too?

Reply:
I guess my question is more about "what is the api for `toward`".  I don't think anybody is or should be calling `toward = null`.    I also don't think removing attributes makes sense for this (or other) components?  I don't think any of the underlying code is expect attributes to be removed.  So, I was wondering more if this should be `set toward(s: string) {`, where we only accept a non-null string here.

(Actually, it looks like `toward` must be an array of two elements, each of which is up/down/left/right, if I'm reading the `parseToward` code below?)

Reply:
What the getter should return when `getAttribute` return `null`? Types in getter and setter should match.

Reply:
So, just trying this out, calling `removeAttribute` does not invoke the `attributeChangedCallback` and just sets the attribute to its default(?) value.  This seems like a browser bug, as the documentation says that it should call the `attributeChangedCallback`.  It seems like this would probably will cause issues for things that need to parse the attributes to set other attributes indirectly.

I guess the options are:
```
// This is what Maiko did in resourcebar.ts:

set toward(s: string | null) {
  this.setAttribute('toward', s ?? 'right down');
}
get toward(): string | null {
  return this.getAttribute('toward');
}

// ... OR ... ?

set toward(s: string) {
  this.setAttribute('toward', s);
}
get toward(): string{
  // maybe print an error here if this happens?
  return this.getAttribute('toward') ?? '';
}
```

Or something else?

Reply:
I think the below one works

Reply:
Ok, sorry for the runaround on this.  I was looking at the revert of @MaikoTan's patch and @panicstevenson's fix in #2633.  I think I misunderstood how these getters are working, and didn't think they were getting used anywhere that would return null, but they are getting used in the attaching function when the component is being set up.  So, having them return null when unset is important.  I also think null is better for unset values than empty string, since we have some string attributes.  And, I agree with the previous idea that getters and setters should have the same type.

Now that I understand this better, I think @TRim21's original proposal of:
```typescript
  set toward(s: string | null) {
    if (s === null)
      this.removeAttribute('toward');
    else
      this.setAttribute('toward', s);
```

...is more consistent and better overall.  Sorry about my confusion.

Reply:
ok